src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -748,81 +754,90 @@ private class BrokerTransaction : IRSessionSwitchBrokerTransaction, IRSessionRec
+                    var host = _session._host;
+                    var hostRunTask = _session._hostRunTask;
+
+                    // host may be null if previous attempts to start it have failed
+                    if (host != null) {
+                        // Detach RHost from RSession
+                        host.DetachCallback();
+
+                        // Cancel all current requests
+                        await _session.CancelAllAsync(cancellationToken);
+                    }
 
-                // Start new RHost
-                await _session.StartHostAsyncBackground(_hostToSwitch, _lockToken, cancellationToken);
+                    // Start new RHost
+                    await _session.StartHostAsyncBackground(_hostToSwitch, _lockToken, cancellationToken);
 
-                // Shut down the old host, gracefully if possible, and wait for old hostRunTask to exit;
-                if (hostRunTask != null) {
-                    await StopHostAsync(brokerClient, startupInfo?.Name, host, hostRunTask);
-                }
-                host?.Dispose();
+                    // Shut down the old host, gracefully if possible, and wait for old hostRunTask to exit;
+                    if (hostRunTask != null) {
+                        await StopHostAsync(brokerClient, startupInfo?.Name, host, hostRunTask);
+                    }
+                    host?.Dispose();
 
-                if (hostRunTask != null) {
-                    await hostRunTask;
+                    if (hostRunTask != null) {
+                        await hostRunTask;
+                    }
+                } catch (InstanceDisposedException<RSession> ex) when (ex.Instance == _session) { 
+                } finally {
+                    _hostToSwitch = null;
                 }
-
-                _hostToSwitch = null;
             }
 
             public async Task ReconnectAsync(CancellationToken cancellationToken, ReentrancyToken reentrancyToken) {
+                if (_session._disposeToken.IsDisposed) {
+                    return;
+                }
+
                 if (_lockToken == null) {
                     throw new InvalidOperationException($"{nameof(AcquireLockAsync)} must be called before {nameof(ReconnectAsync)}");
                 }
 
-                _session._disposeToken.ThrowIfDisposed();
-
                 if (_session._startupInfo == null) {
                     // Session never started. No need to restart it.
                     // Reset _initializationLock so that next awaiter can proceed.
                     _lockToken.Reset();
                     return;
                 }
 
-                var host = _session._host;
-                // host may be null if previous attempts to start it have failed
-                if (host != null) {
-                    // Detach RHost from RSession
-                    host.DetachCallback();
+                try {
+                    var host = _session._host;
```

Comment:
You mean these 2 lines:
```{csharp}
                        // Detach RHost from RSession
                        host.DetachCallback();

                        // Cancel all current requests (if any)
                        await _session.CancelAllAsync(cancellationToken);
```

---
In reply to: [83918298](https://github.com/Microsoft/RTVS/pull/2529#discussion_r83918298) [](ancestors = 83918298)